pulsar-broker/src/main/java/org/apache/pulsar/broker/service/ServerCnx.java

```diff
@@ -100,6 +102,8 @@ public ServerCnx(BrokerService service) {
         this.producers = new ConcurrentLongHashMap<>(8, 1);
         this.consumers = new ConcurrentLongHashMap<>(8, 1);
         this.replicatorPrefix = service.pulsar().getConfiguration().getReplicatorPrefix();
+        this.nonPersistentMessageSemaphore = new Semaphore(
```

Comment:
All the events from the same connections are coming from same thread. Even when we write, we typically switch to the event loop thread for that connection. 

Given that, the semaphore could just be a regular (non even `volatile` variable` that we check and increment. No need for thread safety.